pkg/cloud/aws/metadata/provider.go

```diff
@@ -0,0 +1,175 @@
+	PathServicesDomain                                       = "services/domain"
+	PathServicesPartition                                    = "services/partition"
+	PathSystem                                               = "system"
+)
+
+// A Provider gives you convenient access to the metadata of your EC2 instance.
+// See also https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instancedata-data-retrieval.html.
+type Provider interface {
+	GetMetadata(ctx context.Context, path string) (string, error)
+}
+
+type NotAvailableError struct{}
+
+func (err NotAvailableError) Error() string {
+	return "Metadata not available"
+}
+
+type Settings struct {
+	// The Host setting allows you to overwrite the endpoint we query metadata from
+	Host string `cfg:"host" default:"169.254.169.254"`
+	// If you are not running on a EC2 instance, you can set Available to false, and we won't try to query a non-existing
+	// endpoint, but instead return a NotAvailableError. Calling code is expected to handle this error and react accordingly.
+	Available bool `cfg:"available" default:"true"`
+}
+
+type provider struct {
+	httpClient http.Client
+	clock      clock.Clock
+	settings   Settings
+
+	lck            sync.Mutex
+	token          string
+	tokenExpiresAt time.Time
+}
+
+func ProvideProvider(ctx context.Context, config cfg.Config, logger log.Logger) (Provider, error) {
+	return appctx.Provide(ctx, provider{}, func() (Provider, error) {
+		return newProvider(ctx, config, logger, "default")
+	})
+}
+
+func newProvider(ctx context.Context, config cfg.Config, logger log.Logger, name string) (Provider, error) {
+	httpClient, err := http.ProvideHttpClient(ctx, config, logger, "metadata-provider")
+	if err != nil {
+		return nil, fmt.Errorf("failed to create http client: %w", err)
+	}
+
+	var settings Settings
+	config.UnmarshalKey(fmt.Sprintf("cloud.aws.%s.metadata", name), &settings)
+
+	return NewProviderWithInterfaces(httpClient, clock.Provider, settings), nil
+}
+
+func NewProviderWithInterfaces(httpClient http.Client, clock clock.Clock, settings Settings) Provider {
+	return &provider{
+		httpClient: httpClient,
+		clock:      clock,
+		settings:   settings,
+	}
+}
+
+func (m *provider) GetMetadata(ctx context.Context, path string) (string, error) {
+	if !m.settings.Available {
+		return "", NotAvailableError{}
+	}
+
+	token, err := m.getToken(ctx)
+	if err != nil {
+		return "", fmt.Errorf("failed to get new token: %w", err)
+	}
+
+	req := m.httpClient.NewRequest().
+		WithUrl(fmt.Sprintf("http://%s/latest/meta-data/%s", m.settings.Host, path)).
+		WithHeader("X-aws-ec2-metadata-token", token)
+	res, err := m.httpClient.Get(ctx, req)
+	if err != nil {
+		return "", err
+	}
+
+	if res.StatusCode != netHttp.StatusOK {
+		return "", fmt.Errorf("unexpectd status %d", res.StatusCode)
+	}
+
+	return string(res.Body), nil
+}
+
+func (m *provider) getToken(ctx context.Context) (string, error) {
+	m.lck.Lock()
```

Comment:
Nit: sync.RWMutex with early return would be more elegant

Reply:
I disagree. For the simple case that we already have a token, locking just a normal mutex is enough. There are no concurrent modifications happening until our token expires.

However, if our token is expired / was never set before, we now have really ugly logic: We now need to RLock, then check if the token is acceptable, if not, we have to RUnlock, Lock, check the token again (because we can't upgrade our lock and someone else could've fetched it after we RUnlock'ed) and then do the fetch (unless we want multiple fetches at the same time, which I would prefer to avoid).

All these problems disappear by using a normal mutex - because we have no place where we just read the token, we always have the chance to write it as well (because there is just a single place we need the lock for).

Reply:
Fair argument, conversation resolved :+1: